crates/astria-sequencer/src/proposal/commitment.rs

```diff
@@ -0,0 +1,190 @@
+use std::collections::BTreeMap;
+
+use bytes::Bytes;
+use tendermint::merkle::simple_hash_from_byte_vectors;
+
+use crate::transaction::Signed;
+
+/// Called when we receive a `PrepareProposal` or `ProcessProposal` consensus message.
+///
+/// In the case of `PrepareProposal`, we use this function to generate the `commitment_tx`
+/// to be placed at the start of the block.
+///
+/// In the case of `ProcessProposal`, we use this function to generate and verify the
+/// `commitment_tx` expected at the start of the block.
+///
+/// This function sorts the block's `sequence::Action`s contained within the transactions
+/// using their `chain_id`. It then returns the merkle root of the tree where each leaf is
+/// a commitment of `sequence::Action`s with the same `chain_id`. The leaves are ordered
+/// by namespace in ascending order, where `namespace(chain_id) = Sha256(chain_id)[0..10]`.
+/// This structure can be referred to as the "action tree".
+///
+/// The leaf, which contains a commitment to every action with the same `chain_id`, is currently
+/// implemented as ( `namespace(chain_id)` || root of merkle tree of the `sequence::Action`s ).
+/// This is somewhat arbitrary, but could be useful for proof of an action within the action tree.
+pub(crate) fn generate_sequence_actions_commitment(txs_bytes: &[Bytes]) -> [u8; 32] {
+    // ignore any transactions that are not deserializable
+    let txs = txs_bytes
+        .iter()
+        .filter_map(|tx_bytes| Signed::try_from_slice(tx_bytes).ok())
```

Comment:
Now I am worried about us silently dropping transactions that could not be read. In a first step, can you log them?

But more to the point, I am reading this https://docs.cometbft.com/v0.37/spec/abci/abci++_methods#prepareproposal

And it notes that applications are allowed to remove txs from a proposal and instead put an updated list into `ResponsePrepareProposal.txs`. Should we follow this strategy here?

I am not totally sure if this is the intended use, but it seems reasonable?